qiskit/transpiler/passes/basis/basis_translator.py

```diff
@@ -77,19 +81,37 @@ def run(self, dag):
         Returns:
             DAGCircuit: translated circuit.
         """
-
-        if self._target_basis is None:
+        if self._target_basis is None and self._target is None:
             return dag
 
+        qarg_indices = {qubit: index for index, qubit in enumerate(dag.qubits)}
         # Names of instructions assumed to supported by any backend.
-        basic_instrs = ["measure", "reset", "barrier", "snapshot", "delay"]
-
-        target_basis = set(self._target_basis).union(basic_instrs)
+        if self._target is None:
+            basic_instrs = ["measure", "reset", "barrier", "snapshot", "delay"]
+            target_basis = set(self._target_basis)
+            source_basis = set()
+            for node in dag.op_nodes():
+                if not dag.has_calibration_for(node):
+                    source_basis.add((node.name, node.op.num_qubits))
+            qarg_with_incomplete = {}
+            incomplete_source_basis = {}
+        else:
+            basic_instrs = ["barrier", "snapshot"]
+            source_basis = set()
+            target_basis = self._target.keys() - set(self._incomplete_basis)
+            qarg_with_incomplete = defaultdict(set)
+            for gate in self._incomplete_basis:
+                for qarg in self._target[gate]:
+                    qarg_with_incomplete[qarg].add(gate)
+            incomplete_source_basis = defaultdict(set)
+            for node in dag.op_nodes():
+                qargs = tuple(qarg_indices[bit] for bit in node.qargs)
```

Comment:
Do we also need an `if not dag.has_calibration_for(node):` here?

Reply:
Done in: https://github.com/Qiskit/qiskit-terra/pull/7227/commits/d2c1e11b894420e1a44b120e1ba6b13a1a87a26f